Bug 2041765: Adjust the startup order of httpd container

zhouhao3 opened the pull request:
Run the httpd container after the coreos-downloader completes to ensure that the kernel parameters can be added correctly.

Fixes: #5504 

Signed-off-by: Zhou Hao <zhouhao@fujitsu.com>

rhjanders commented:
/ok-to-test

rhjanders commented:
/retest

zaneb commented:
So the runhttpd script is [rendering](https://github.com/openshift/ironic-image/blob/master/scripts/runhttpd#L33) `inspector.ipxe`, which [contains](https://github.com/openshift/ironic-image/blob/master/ironic-config/inspector.ipxe.j2#L8) a reference to `IRONIC_KERNEL_PARAMS`, which is only set correctly [if the rootfs file exists](https://github.com/openshift/ironic-image/blob/master/scripts/coreos-ipa-common.sh#L15-L21).

What's most confusing about this is how it could have been working anywhere?

We'll need to link this to a (public) bugzilla in order to merge it.

/approve
/label platform/baremetal
/test e2e-metal-ipi

zhouhao3 commented:
> What's most confusing about this is how it could have been working anywhere?

We are also very confused, what environment do you test in so that this issue will not occur? Maybe it's just an iPXE issue.

rhjanders commented:
/test e2e-aws-workers-rhel8

rhjanders commented:
/test e2e-metal-ipi-ovn-dualstack

zhouhao3 commented:
/retitle Bug 2041765: Adjust the startup order of httpd container

zhouhao3 commented:
> We'll need to link this to a (public) bugzilla in order to merge it.

Sent bugzilla https://bugzilla.redhat.com/show_bug.cgi?id=2041765

zaneb commented:
/bugzilla refresh

derekhiggins commented:
> What's most confusing about this is how it could have been working anywhere?

AIUI there are two paths to inspecting a host, requesting ironic-conducor to inspect a specific host will cause ironic to create a pxe config specific for the host, this uses dualboot.ipxe this doesn't not use inspector.ipxe as a config is present for the MAC of the host

The alternative also uses dualboot.ipxe but fails to get a host specific config and then fallsback to inspector.ipxe , this could happen if
 inspection is not being managed by ironic-conductor, perhaps by talking directly to inspector?
or the MAC address for the host is incorrect, causing the config created for the host to fail to be found

zhouhao3 commented:
>The alternative also uses dualboot.ipxe but fails to get a host specific config and then fallsback to inspector.ipxe , this could happen if
inspection is not being managed by ironic-conductor, perhaps by talking directly to inspector?
or the MAC address for the host is incorrect, causing the config created for the host to fail to be found

As we can see from the PXE startup interface, inspector.ipxe is used because there is no corresponding file in the pxelinux.cfg folder.

![MicrosoftTeams-image](https://user-images.githubusercontent.com/19640187/150073067-9f19be85-3358-4f6b-adaa-be5d21b9aa91.png)

We can see that the pxelinux.cfg folder is empty in bootstrap.

```
[root@bootstrap html]# pwd
/var/lib/containers/storage/volumes/ironic/_data/html
[root@bootstrap html]# ls
boot.ipxe  dualboot.ipxe  images  inspector.ipxe  pxelinux.cfg  uefi_esp.img
[root@bootstrap html]# ls pxelinux.cfg/
```

There are relevant files in the tftpboot folder, which are produced by the dnsmasq container.

```
[root@bootstrap _data]# pwd
/var/lib/containers/storage/volumes/ironic/_data
[root@bootstrap _data]# ls
html  ironic_prometheus_exporter  log  tftpboot
[root@bootstrap _data]# ls tftpboot/pxelinux.cfg/
01-90-1b-0e-e6-37-dd  01-90-1b-0e-e6-7b-70  01-90-1b-0e-e6-7b-88
[root@bootstrap _data]# pwd
/var/lib/containers/storage/volumes/ironic/_data
[root@bootstrap _data]# ls
html  ironic_prometheus_exporter  log  tftpboot
[root@bootstrap _data]# ls tftpboot/
01e9193d-b574-4daa-922b-96204e9b0ca6            90:1b:0e:e6:37:dd.conf                          grub.cfg-01-90-1b-0e-e6-37-dd  snponly.efi
19c783cc-80f1-56d9-b495-de809b584454.converted  90:1b:0e:e6:7b:70.conf                          grub.cfg-01-90-1b-0e-e6-7b-70  undionly.kpxe
207cca95-4faf-5c21-b7ac-05c88802adec.converted  90:1b:0e:e6:7b:88.conf                          grub.cfg-01-90-1b-0e-e6-7b-88
5dcaec5b-7e86-5ba6-849b-5d554febc5fc.converted  bb03b882-cb00-41ad-bf24-1dd6dd6fc8a9            ipxe.efi
80c6b42d-fc31-42c9-b712-0d968f261488            efab1372-ace5-5041-922e-851381838617.converted  pxelinux.cfg
[root@bootstrap _data]# ls -l tftpboot/pxelinux.cfg/
total 0
lrwxrwxrwx. 1 root root 46 Jan 19 05:47 01-90-1b-0e-e6-37-dd -> ../80c6b42d-fc31-42c9-b712-0d968f261488/config
lrwxrwxrwx. 1 root root 46 Jan 19 05:47 01-90-1b-0e-e6-7b-70 -> ../01e9193d-b574-4daa-922b-96204e9b0ca6/config
lrwxrwxrwx. 1 root root 46 Jan 19 05:47 01-90-1b-0e-e6-7b-88 -> ../bb03b882-cb00-41ad-bf24-1dd6dd6fc8a9/config

```
It looks like the tftp service configured in the dnsmasq container is correct, but it is not being used.

zaneb commented:
The files in the tftpboot dir have `01-` prepended to the MAC, which happens when `ipxe_enabled` is false, but the dualboot.ipxe script is just looking for the MAC. So it sounds to me like we are somehow using ipxe but with `ipxe_enabled = False` in the irmc driver/config somehow.

We are setting the boot interface to `pxe`, so that is probably wrong: https://github.com/metal3-io/baremetal-operator/blob/main/pkg/hardwareutils/bmc/irmc.go#L77-L79

According to https://docs.openstack.org/ironic/latest/admin/drivers/irmc.html#hardware-interfaces the `pxe` interface is deprecated and we should use `irmc-pxe`.
https://docs.openstack.org/ironic/latest/admin/drivers/irmc.html#boot-from-remote-volume says that both of those use iPXE, but that does not appear to be true - `pxe` and `ipxe` are different boot interfaces, and `IRMCPXEBoot` inherits from `pxe.PXEBoot` (with `ipxe_enabled = False), not `ipxe.iPXEBoot`.

I suspect if we change the boot interface to `ipxe`, this will work.

zhouhao3 commented:
>I suspect if we change the boot interface to ipxe, this will work.

After changing the boot interface to ipxe, there will be no previous issue, thank you very much, I will close this PR, and I will create a fixed PR later.

zhouhao3 commented:
After solving this problem by changing the boot interface to ipxe, there is a issue that the Master cannot download the Ironic Agent image. See #5552 for details.

zaneb commented:
I think we should still merge this, since it's clearly correct even if it isn't strictly needed in this case.
/reopen

rhjanders commented:
/ok-to-test

derekhiggins commented:
/lgtm

rhjanders commented:
/test e2e-metal-single-node-live-iso

rhjanders commented:
/test e2e-metal-single-node-live-iso

rhjanders commented:
/test e2e-aws-workers-rhel8

ardaguclu commented:
/approve

patrickdillon commented:
alibaba should no longer be required.

/skip
